test_util/src/testenv.rs

```diff
@@ -0,0 +1,342 @@
+    env, fs,
+    io::{self, Write},
+    path::{Path, PathBuf},
+    sync::{Arc, Mutex, PoisonError},
+};
+use tempfile::{tempdir, TempDir};
+
+/// Env var convenience type.
+type EnvMap = HashMap<String, String>;
+
+/// Initialized in [`get_env_locker`]
+static ENV_LOCKER: OnceCell<Arc<Mutex<EnvMap>>> = OnceCell::new();
+
+/// A test environment.
+///
+/// Will create a new temporary directory. Use its builder methods to configure
+/// the directory structure, preset variables, envfile name and contents, and
+/// the working directory to run the test from.
+///
+/// Creation methods:
+/// - [`TestEnv::init`]: blank environment (no envfile)
+/// - [`TestEnv::init_with_envfile`]: blank environment with a custom `.env`
+/// - [`TestEnv::default`]: default testing environment (1 existing var and 2
+///       set in a `.env` file)
+#[derive(Debug)]
+pub struct TestEnv {
+    // Temporary directory that will be deleted on drop
+    _temp_dir: TempDir,
+    dir_path: PathBuf,
+    work_dir: PathBuf,
+    env_vars: EnvMap,
+    envfiles: Vec<EnvFile>,
+}
+
+#[derive(Debug, Clone)]
+/// Simple path and byte contents representing a `.env` file
+pub struct EnvFile {
+    pub path: PathBuf,
+    pub contents: Vec<u8>,
+}
+
+/// Run a test closure within a test environment.
+///
+/// Resets the environment variables, loads the [`TestEnv`], then runs the test
+/// closure. Ensures only one thread has access to the process environment.
+pub fn test_in_env<F>(testenv: &TestEnv, test: F)
+where
+    F: FnOnce(),
+{
+    let locker = get_env_locker();
+    // ignore a poisoned mutex
+    // we expect some tests may panic to indicate a failure
+    let original_env = locker.lock().unwrap_or_else(PoisonError::into_inner);
+    // we reset the environment anyway upon acquiring the lock
+    reset_env(&original_env);
+    create_env(testenv);
+    test();
+    // drop the lock
+}
+
+/// Run a test closure within the default test environment.
+///
+/// Resets the environment variables, creates the default [`TestEnv`], then runs
+/// the test closure. Ensures only one thread has access to the process
+/// environment.
+///
+/// The default testing environment sets an existing environment variable
+/// `DEFAULT_EXISTING_KEY`, which is set to `loaded_from_env`. It also creates a
+/// `.env` file with the two lines:
+///
+/// ```ini
+/// DEFAULT_TEST_KEY=default_test_val
+/// DEFAULT_EXISTING_KEY=loaded_from_file
+/// ```
+///
+/// Notice that file has the potential to override `DEFAULT_EXISTING_KEY` depending
+/// on the what's being tested.
+pub fn test_in_default_env<F>(test: F)
+where
+    F: FnOnce(),
+{
+    let testenv = TestEnv::default();
+    test_in_env(&testenv, test);
+}
+
+/// Create a [`TestEnv`] without an envfile, but with the
+/// default existing environment variable.
+pub fn create_testenv_with_default_var() -> TestEnv {
+    let mut testenv = TestEnv::init();
+    testenv.add_env_var(DEFAULT_EXISTING_KEY, DEFAULT_EXISTING_VALUE);
+    testenv
+}
```

Comment:
This should be impl on `TestEnv`. Would it be detrimental to tests if we had this set as default for `TestEnv`, such as in `TestEnv::init`?

Reply:
Yes, I think simpler is better!